core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/CostBasedRegionPlanGenerator.scala

```diff
@@ -326,6 +330,125 @@ class CostBasedRegionPlanGenerator(
+    * will be performed to find an optimal plan. The search starts from a plan where all edges are materialized, and
+    * leads to a low-cost schedulable plan by changing materialized non-blocking edges to pipelined.
+    * By default, all pruning techniques are enabled (chains, clean edges).
+    *
+    * @return A SearchResult containing the plan, the region DAG (without materializations added yet), the cost, the
+    *         time to finish search, and the number of states explored.
+    */
+  def topDownSearch(
+      globalSearch: Boolean = false,
+      oChains: Boolean = true,
+      oCleanEdges: Boolean = true
+  ): SearchResult = {
+    val startTime = System.nanoTime()
+    // Starting from a state where all non-blocking edges are materialized
+    val originalSeedState = physicalPlan.links.diff(
+      physicalPlan.getNonMaterializedBlockingAndDependeeLinks
+    )
+
+    // Chain optimization: an edge in the same chain as a blocking edge should not be materialized
+    val seedStateOptimizedByChainsIfApplicable = if (oChains) {
+      val edgesInChainWithBlockingEdge = physicalPlan.maxChains
+        .filter(chain =>
+          chain.intersect(physicalPlan.getNonMaterializedBlockingAndDependeeLinks).nonEmpty
+        )
+        .flatten
+      originalSeedState.diff(edgesInChainWithBlockingEdge)
+    } else {
+      originalSeedState
+    }
+
+    // Clean edge optimization: a clean edge should not be materialized
+    val finalSeedState = if (oCleanEdges) {
+      seedStateOptimizedByChainsIfApplicable.intersect(physicalPlan.getNonBridgeNonBlockingLinks)
+    } else {
+      seedStateOptimizedByChainsIfApplicable
+    }
+
+    // Queue to hold states to be explored, starting with the seed state
+    val queue: mutable.Queue[Set[PhysicalLink]] = mutable.Queue(finalSeedState)
+    // Keep track of visited states to avoid revisiting
+    val visited: mutable.Set[Set[PhysicalLink]] = mutable.Set.empty[Set[PhysicalLink]]
+    // Initialize the bestResult with an impossible high cost for comparison
+    var bestResult: SearchResult = SearchResult(
+      state = Set.empty,
+      regionDAG = new DirectedAcyclicGraph[Region, RegionLink](classOf[RegionLink]),
+      cost = Double.PositiveInfinity
+    )
+
+    while (queue.nonEmpty) {
+      val currentState = queue.dequeue()
+      visited.add(currentState)
+      tryConnectRegionDAG(
+        physicalPlan.getNonMaterializedBlockingAndDependeeLinks ++ currentState
+      ) match {
+        case Left(regionDAG) =>
+          updateOptimumIfApplicable(regionDAG)
+          addNeighborStatesToFrontier()
+        // No need to explore further
+        case Right(_) =>
+          addNeighborStatesToFrontier()
+      }
+
+      /**
+        * An internal method of top-down search that updates the current optimum if the examined state is schedulable
+        * and has a lower cost.
+        */
+      def updateOptimumIfApplicable(regionDAG: DirectedAcyclicGraph[Region, RegionLink]): Unit = {
```

Comment:
this method can be used by bottom up as well right?

Reply:
No, because it needs to access the local variable `bestResult` of the top-down search method.

Reply:
ok